architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,550 @@
+	BicepDeploymentPhraseDeleting BicepDeploymentPhrase = "Deleting"
+
+	// BicepDeploymentPhraseDeleted indicates that the Bicep Deployment has been deleted.
+	BicepDeploymentPhraseDeleted BicepDeploymentPhrase = "Deleted"
+)
+
+// BicepDeploymentStatus defines the observed state of the Bicep Deployment.
+type BicepDeploymentStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Bicep Deployment. It corresponds to the Deployment's generation, which is updated on mutation by the API Server.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
+
+	// Scope is the resource ID of the scope.
+	Scope string `json:"scope,omitempty"`
+
+	// Resource is the resource ID of the deployment.
+	Resource string `json:"resource,omitempty"`
+
+	// Operation tracks the status of an in-progress provisioning operation.
+	Operation *ResourceOperation `json:"operation,omitempty"`
+
+	// Phrase indicates the current status of the Bicep Deployment.
+	Phrase BicepDeploymentPhrase `json:"phrase,omitempty"`
+
+  // Description is a human-readable description of the status of the Bicep Deployment.
+  Description string `json:"description,omitempty"`
+
+  // Resources is a list of resources that are created by the BicepDeployment.
+  Resources []BicepDeploymentResource `json:"resources,omitempty"`
+}
+```
+
+#### `rad bicep generate-kubernetes` Command
+
+The `rad bicep generate-kubernetes` command will be a new command that will generate a Kubernetes custom resource when provided a Bicep template, modules, and parameters The command will take the following arguments:
+
+- `-p, --parameters`: The parameters for the Bicep template. Can be specified in the same way as [`rad deploy` parameters](https://edge.docs.radapp.io/reference/cli/rad_deploy/).
+- `-o, --outfile`: The path to the output file where the BicepDeployment resource will be written.
+
+#### BicepDeployment Controller
+
+The `radius-controller` will be updated to include a new controller that reconciles BicepDeployment resources. It will have the following control loop:
+
+ The BicepDeployment controller will adhere to the following control loop for BicepDeployment resources on the cluster:
+
+1. Check if there is an in-progress operation. If so, check its status:
+    1. If the operation is still in progress, then queue another reconcile operation and continue processing.
+	  2. If the operation completed successfully, then update the `status.phrase` and `status.description` fields as succeeded. Set the contents of the `status.resources` to match the set of deployed resources. This operation is complete, so we can continue processing.
+	  3. If the operation failed, then update the `status.phrase` and `status.description` as failed and continue processing.
+2. If the BicepDeployment is being deleted, then process deletion:
+    1. Queue a DELETE operation for each of the resources in the `status.resources` list.
```

Comment:
Let's think about each of the "tracked resources" as their own resources so that they can manage their own async lifecycle.